### ISSUE TITLE
Loss curve changes when loading checkpoint

### ISSUE THREAD

**Issue by okoge-kaz** [microsoft/Megatron-DeepSpeed#248]
## About

I am using Megatron-DeepSpeed after rebase, and comparing the loss curves when training is restarted from checkpoint and when training is started from the beginning without using checkpoint, there are differences.

This loss curve change does not occur in Megatron-DeepSpeed before rebasing

This problem can be critical for large-scale trainings and I would like to resolve it. I would very much appreciate your help.

## Detail

The loss curves are different when learning from scratch without checkpoint and when learning from the middle of the training process using checkpoint, even if the values are consistent using --seed.

There are also differences in the optimizer state.

Below is the data of the experiment with Megatron-DeepSpeed after rebasing.

training loss
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/025f77cd-7170-4914-9c3d-254a78d2c2ca)

training loss (Enlarged version)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/b7a5d32c-55d5-4496-bd59-1c5dbca87ad0)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/b910a79c-e2a5-4e6d-968e-bc3e80d3c84f)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/2d04e4c4-ec71-4de2-a28f-777925704069)

Looking at the loss curve of this Weights & Biases, it appears that a small change occurs at the next iteration after the checkpoint is read. 

You can see the results of my experiment here.
https://wandb.ai/okoge/megatron-deepspeed-loss-spike/reports/loss-curves-are-different---Vmlldzo1Mzk1Mzc4

(full version: https://wandb.ai/okoge/megatron-deepspeed-loss-spike?workspace)

### Experiment environment

 DGX-A100 (40GB) 

before rebase: deepspeed version 0.9.5
after rebase: deepspeed version 0.10.0

### When experimenting with 1 node 1 GPU (after rebase) 

training loss
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/613fe8c0-947d-44ec-8ad9-7c03d8aea9de)

training loss (Enlarged version)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/3530f4cc-42bf-466e-a973-d5c132bbb18e)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/dd019ee9-42ca-4178-9883-c58751bc6d76)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/d1f5bbcb-f245-4837-809e-f051a886ab40)

From these results, it seems that the Optimizer state is not loading well.

###  1 node 8 GPU DGX-A100 (40GB) without `--deepspeed` ( before rebase )

training loss
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/bcf3091f-f168-4baa-8b8f-5e98a0f3ece1)

training loss (Enlarged version)
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/d3c82736-ca86-46e8-9e87-81ed114fb597)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/95430b2d-1108-484b-a10e-f64ecf1d9002)

This problem does not seem to have occurred with Megatron-DeepSpeed before the rebase

wandb report: https://wandb.ai/okoge/megatron-deepspeed-loss-spike/reports/Megatron-DeepSpeed-without-deepspeed-before-rebase--Vmlldzo1Mzk1NTg1

### 1 node 8 GPU  DGX-A100 (40GB) with`--deepspeed` ( before rebase )

training loss
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/b63cc103-d400-4607-813b-e5408d31da78)

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/f7d59fb3-85ee-4ed4-ada1-bf9aaf224103)

grad-norm
![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/3ea5a2d0-5f32-40fb-b45a-ae60268afc2b)

The curve of loss and the transition of grad-norm have not changed depending on whether CHECKPOINT was used or not.

**Comment by okoge-kaz** [comment]
Here is a link to the google slide that I kept as a note during my investigation. (Sorry it is mainly in Japanese.)

https://docs.google.com/presentation/d/1NstfuGmxfhXDcMPBw2gT3XQ215WmnZuDRHabr2_nIc0/edit?usp=sharing

**Comment by conglongli** [comment]
The scripts and ds config for reproducing the above issue from @okoge-kaz :
1 node 8 gpu: https://github.com/llm-jp/Megatron-DeepSpeed/blob/other/fujii/loss-spike-investigation/scripts/mpirun/1.3B/ylab/1.3B_dp8_tp1_pp1_zero1.sh
1 node 1 gpu: https://github.com/llm-jp/Megatron-DeepSpeed/blob/other/fujii/loss-spike-investigation/scripts/mpirun/350m/350m_dp1_tp1_pp1_zero.sh

ds_config.json: https://github.com/llm-jp/Megatron-DeepSpeed/blob/other/fujii/loss-spike-investigation/examples_deepspeed/rebase/ds_config_gpt_TEMPLATE.json

Custom tokenizker data was used, but potentially we can first try to reproduce the issue with the default GPT2Tokenizer and data @tjruwase @tohtana.

**Comment by okoge-kaz** [comment]
The job script used in the experiment in before rebase version.

1 node 8 gpu: https://github.com/llm-jp/Megatron-DeepSpeed/blob/other/fujii/before-rebase-loss-spike/scripts/1.3B_dp8_tp1_pp1-fp16-ja90_en10_deepspeed.sh

ds_config.json: https://github.com/llm-jp/Megatron-DeepSpeed/blob/other/fujii/before-rebase-loss-spike/scripts/config-dgx.json

**Comment by okoge-kaz** [comment]
I can check that the problem has been resolved!
Thank you all for your help!

![image](https://github.com/microsoft/Megatron-DeepSpeed/assets/68278821/9901a321-0c26-4329-82a2-8bca1c9ffd18)
